lib/adapters/class-vip-enterprise-search.php

```diff
@@ -197,6 +255,168 @@ public function filter__vip_search_post_taxonomies_allow_list( $taxonomies ) {
+	 *     @type int[]    $include  Search within these post IDs.
+	 *     @type int[]    $exclude  Exclude these post IDs from results.
+	 *     @type string[] $status   Post statuses to search. Default 'publish'.
+	 * }
+	 * @return int[] Post IDs in this page of results and total number of results.
+	 */
+	public function query_post_suggestions( string $search, array $args = [] ): array {
+		$out = [ [], 0 ];
+
+		if ( ! $search && ! $this->get_allow_empty_search() ) {
+			return $out;
+		}
+
+		$args = wp_parse_args(
+			$args,
+			[
+				'subtypes' => [],
+				'page'     => 1,
+				'per_page' => 10,
+				'exclude'  => [], // phpcs:ignore WordPressVIPMinimum.Performance.WPQueryParams.PostNotIn_exclude
+				'include'  => [],
+				'status'   => [ 'publish' ],
+			],
+		);
+
+		$subtypes = (array) $args['subtypes'];
+		$page     = (int) $args['page'];
+		$per_page = (int) $args['per_page'];
+		$exclude  = (array) $args['exclude'];
+		$include  = (array) $args['include'];
+
+		if ( $search ) {
+			$query_from     = ( $page - 1 ) * $per_page;
+			$query_per_page = $per_page;
+			$query_must     = [
+				[
+					'multi_match' => [
+						'query'    => $search,
+						'type'     => 'bool_prefix',
+						'fields'   => [
+							'search_suggest',
+							'search_suggest._2gram',
+							'search_suggest._3gram',
+						],
+
+						/*
+						 * The purpose of using 'and' here is to assume that more terms in the search indicate
+						 * someone searching for something more specific, not someone trying to get more results.
+						 */
+						'operator' => 'and',
+						'_name'    => 'search',
+					],
+				],
+				[
+					'terms' => [
+						'post_status' => (array) $args['status'],
+						'_name'       => 'status',
+					],
+				],
+			];
+			$query_must_not = [];
+
+			if ( $subtypes ) {
+				$restrict = $this->get_restricted_search_suggestions_post_types();
+
+				/*
+				 * If search suggestions have been limited to a list of post types,
+				 * don't suggest posts from any other types, even if posts in other
+				 * types have suggestions indexed because they used to be allowed.
+				 */
+				if ( $restrict ) {
+					$subtypes = array_intersect( $subtypes, $restrict );
+
+					if ( ! $subtypes ) {
+						$subtypes = $restrict;
+					}
+				}
+
+				$query_must[] = [
+					'terms' => [
+						'post_type.raw' => $subtypes,
+						'_name'         => 'subtypes',
+					],
+				];
+			}
+
+			if ( $include ) {
+				$query_must[] = [
+					'terms' => [
+						'post_id' => $include,
+						'_name'   => 'include',
+					],
+				];
+			}
+
+			if ( $exclude ) {
+				$query_must_not[] = [
+					'terms' => [
+						'post_id' => $exclude,
+						'_name'   => 'exclude',
+					],
+				];
+			}
+
+			$query = [
+				'from'    => $query_from,
+				'size'    => $query_per_page,
+				'query'   => [
+					'bool' => [
+						'must'     => $query_must,
+						'must_not' => $query_must_not,
+					],
+				],
+				'_source' => [
+					'ID',
+				],
+			];
```

Comment:
🤔  Should we add a hook in case it might be needed?

Reply:
I was hesitant to do so because I think it would have to be a filter that was specific to the VIP Search implementation, wouldn't it? It couldn't be a generic filter, like `elasticsearch_extensions_whatever`, because the DSL passed to the filter might change with each provider. You couldn't be sure that your filter would continue to work if you switched from SearchPress to VIP Search, for instance. What do you think?

Reply:
That's a very good point. I think this raises bigger questions, particularly about switching adapters. Better not adding a hook filter then.